scipy/special/xsf/tools.h

```diff
@@ -274,10 +271,163 @@ namespace detail {
+            frontier = x0 + step0_left;
+            y_interior_sgn = y0_sgn;
+            search_left = true;
+            boundary = xmin;
+            factor = factor_left;
+        }
+
+        bool reached_boundary = false;
+        for (std::uint64_t i = 0; i < maxiter; i++) {
+            y_frontier = func(frontier);
+            y_frontier_sgn = std::signbit(y_frontier);
+            if (y_frontier_sgn != y_interior_sgn || (y_frontier == 0.0)) {
+                /* Stopping condition, func evaluated at endpoints of bracket has opposing signs,
+                 * meeting requirement for bracketing root finder. (Or endpoint has reached a
+                 * zero.) */
+                if (search_left) {
+                    /* Ensure we return an interval (a, b) with a < b. */
+                    std::swap(interior, frontier);
+                    std::swap(y_interior, y_frontier);
+                }
+		return {interior, frontier, y_interior, y_frontier, 0};
+            }
+            if (reached_boundary) {
+                /* We've reached a boundary point without finding a root . */
+                return {
+                    std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN(),
+                    std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN(),
+                    search_left ? 1 : 2
+                };
+            }
+            double step = (frontier - interior) * factor;
+            interior = frontier;
+            y_interior = y_frontier;
+            y_interior_sgn = y_frontier_sgn;
+            frontier += step;
+            if ((search_left && frontier <= boundary) || (!search_left && frontier >= boundary)) {
+                /* If the frontier has reached the boundary, set a flag so the algorithm will know
+                 * not to search beyond this point. */
+                frontier = boundary;
+                reached_boundary = true;
+            }
+        }
+        /* Failed to converge within maxiter iterations. If maxiter is sufficiently high and
+         * factor_left and factor_right are set appropriately, this should only happen due to
+         * a bug in this function. Limiting the number of iterations is a defensive programming measure. */
+	return {
+            std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN(),
+            std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN(), 3
+        };
+    }
+
+    /* Find root of a scalar function using Chandrupatla's algorithm */
+    template <typename Function>
+    XSF_HOST_DEVICE inline std::pair<double, int> find_root_chandrupatla(
+        Function func, double x1, double x2, double f1, double f2, double rtol,
+        double atol, std::uint64_t maxiter
+    ) {
+        if (f1 == 0) {
+            return {x1, 0};
+        }
+        if (f2 == 0) {
+            return {x2, 0};
+        }
```

Comment:
> * Likewise, an example of "It is expected that the rootfinder will check if the bracket endpoints are roots."

Here are the lines where this happens. If we end up implementing other bracketing root finding algorithms, then we'd want them all to have this property I guess.

The same example I gave before will work here. The initial bracket will be `$[1.0, 1.0]$, and the endpoints get checked.